_drafts/cortex-m-breakpoints.md

```diff
@@ -0,0 +1,607 @@
+Sending packet: $Z0,2a0,2#d7...Packet received: OK
+Sending packet: $Z0,20000000,2#96...Packet received: OK
+Sending packet: $c#63...
+```
+
+```
+shell> fpb_dump
+FPB Revision: 0, Enabled: 1, Hardware Breakpoints: 6
+  FP_COMP[0] Enabled 1, Replace: 1, Address 0x250
+  FP_COMP[1] Enabled 1, Replace: 1, Address 0x260
+  FP_COMP[2] Enabled 1, Replace: 1, Address 0x270
+  FP_COMP[3] Enabled 1, Replace: 1, Address 0x280
+  FP_COMP[4] Enabled 1, Replace: 1, Address 0x290
+  FP_COMP[5] Enabled 1, Replace: 1, Address 0x2a0
+```
+
+Interesting, it looks like the breakpoint for `dummy_function_1` (address 0x240) has been removed from `FP_COMP_0` and `dummy_function_2` through `dummy_function_7` have been configured as hardware breakpoints.
+
+Let's call all the functions, using `call_dummy_funcs` and confirm we break at `dummy_function_1`..`dummy_function_7` as well as `dummy_function_ram`:
+
+```bash
+(gdb) c
+Continuing.
+
+Breakpoint 1, dummy_function_1 () at dummy_functions.c:4
+4     EXAMPLE_LOG("stub function '%s' called", __func__);
+(gdb) c
+Continuing.
+
+Breakpoint 3, dummy_function_2 () at dummy_functions.c:8
+8     EXAMPLE_LOG("stub function '%s' called", __func__);
+(gdb) c
+Continuing.
+
+[...]
+
+Breakpoint 8, dummy_function_7 () at dummy_functions.c:28
+28    EXAMPLE_LOG("stub function '%s' called", __func__);
+(gdb) c
+Continuing.
+
+Breakpoint 2, dummy_function_ram () at dummy_functions.c:41
+41    EXAMPLE_LOG("stub function '%s' called", __func__);
+(gdb) c
+```
+
+Neat, they all worked so a breakpoint was installed for our flash-based function, `dummy_function_1`, that was missing from the FPB configuration.
+
+Let's dump the functions to examine if any instructions were modified:
+
+```
+shell> dump_dummy_funcs
+dummy_function_1: Starts at 0x240. First Instruction = 0x4802be00
+dummy_function_2: Starts at 0x250. First Instruction = 0x48024901
+dummy_function_3: Starts at 0x260. First Instruction = 0x48024901
+dummy_function_4: Starts at 0x270. First Instruction = 0x48024901
+dummy_function_5: Starts at 0x280. First Instruction = 0x48024901
+dummy_function_6: Starts at 0x290. First Instruction = 0x48024901
+dummy_function_7: Starts at 0x2a0. First Instruction = 0x48024901
+dummy_function_8: Starts at 0x2b0. First Instruction = 0x48024901
+dummy_function_9: Starts at 0x2c0. First Instruction = 0x48024901
+dummy_function_ram: Starts at 0x20000000. First Instruction = 0x4802be00
+```
+
+Aha, we can see `dummy_function_1` also was updated to include a breakpoint instruction, `0xbe00`. What's interesting here is this address is in the NRF52's internal flash space. This means to inject this breakpoint the flash page the address is in must have been reprogrammed.
+
+This is in fact exactly what happened. The SEGGER JLinkGDBServer implements a feature referred to in the docs as "unlimited flash breakpoints"[^11] which will patch flash regions the J-Link Software knows how to write to with `bkpt` instructions.
```

Comment:
Could add a note that says "this means that the JLink software ships with drivers for all those chips since they are not standard".

Reply:
There's also QSPI parts with XIP they support by default too fwiw!